vapi/tags/tags.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tags
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+const (
+	TagURL = "/com/vmware/cis/tagging/tag"
+)
+
+type TagCreateSpec struct {
+	CreateSpec TagCreate `json:"create_spec"`
+}
+
+type TagCreate struct {
+	CategoryID  string `json:"category_id"`
+	Description string `json:"description"`
+	Name        string `json:"name"`
+}
+
+type TagUpdateSpec struct {
+	UpdateSpec TagUpdate `json:"update_spec,omitempty"`
+}
+
+type TagUpdate struct {
+	Description string `json:"description,omitempty"`
+	Name        string `json:"name,omitempty"`
+}
+
+type Tag struct {
+	ID          string   `json:"id"`
+	Description string   `json:"description"`
+	Name        string   `json:"name"`
+	CategoryID  string   `json:"category_id"`
+	UsedBy      []string `json:"used_by"`
+}
+
+func (c *RestClient) CreateTagIfNotExist(ctx context.Context, name string, description string, categoryID string) (*string, error) {
+	tagCreate := TagCreate{categoryID, description, name}
+	spec := TagCreateSpec{tagCreate}
+	id, err := c.CreateTag(ctx, &spec)
+	if err == nil {
+		return id, nil
+	}
+	// if already exists, query back
+	if strings.Contains(err.Error(), ErrAlreadyExists) {
+		tagObjs, err := c.GetTagByNameForCategory(ctx, name, categoryID)
+		if err != nil {
+			return nil, err
+		}
+		if tagObjs != nil {
+			return &tagObjs[0].ID, nil
+		}
+
+		// should not happen
+		return nil, fmt.Errorf("Failed to create tag for it's existed, but could not query back. Please check system")
+	}
+
+	return nil, fmt.Errorf("Created tag failed for %s", err)
+}
+
+func (c *RestClient) DeleteTagIfNoObjectAttached(ctx context.Context, id string) error {
+	objs, err := c.ListAttachedObjects(ctx, id)
+	if err != nil {
+		return err
+	}
+	if objs != nil && len(objs) > 0 {
+		return fmt.Errorf("tag %s related objects is not empty, do not delete it", id)
+	}
+	return c.DeleteTag(ctx, id)
+}
+
+func (c *RestClient) CreateTag(ctx context.Context, spec *TagCreateSpec) (*string, error) {
+	stream, _, status, err := c.call(ctx, "POST", TagURL, spec, nil)
```

Comment:
We could use the `http.MethodPost` constant instead of literal `"POST"`, same for the other http methods.